### PR TITLE
create: questions pages for viewing and managing them 

### PR DESCRIPTION
**Pull request by NBNARADHYA**
### Frontend

1. Create practice page to display all public questions.

practicePage

### Backend

1. Update `POST /questions` endpoint to support offset pagination with page_count being returned on every request.
2. Fix a small bug on `POST /questions` related to `confidential_questions`
2. Change `GET /tag` endpoint to `GET /tags`.
3. Update `GET /tags` endpoint to serve unauthorized users.

### CODE REVIEW

File: server/models/questions/getPublicContestQuestions.js [hunk 1]
```diff
@@ -1,59 +1,133 @@
...
+        difficulty === 'hard')
+    ) {
+      query += `difficulty=? AND `
+      countQuery += `difficulty=? AND `
+      qArr.push(difficulty)
+      cQArr.push(difficulty)
     }
 
     query += `c.end_time <= NOW() `
+    countQuery += `c.end_time <= NOW() `
```
Comment (cjchirag7) [inline server/models/questions/getPublicContestQuestions.js:94]: This endpoint is for practice questions and show return questions of only public contests and contests which do not contain confidential-questions.

```suggestion
    query += `c.end_time <= NOW()  AND c.public  = 1 AND c.confidential_questions = 0 `
    countQuery += `c.end_time <= NOW()  AND c.public = 1 AND c.confidential_questions = 0 `
```

**Review by cjchirag7 — CHANGES_REQUESTED** [review]
A minor change required in the API for GET practice questions, in backend. Rest, backend looks fine. Still reviewing the frontend.

**Comment by cjchirag7** [comment]
I have made the following changes to this pull request-

1. When any search query or tags or page size or sort method changes, then index of page number needs be changed to 1.  
2. Added caret icon in dropdown button title
3. Show description of tag, when hovered over any tag.
4. Extend pagination to include previous & next buttons, along with 5-7 nearest page numbers
5. Highlight `practice` in navbar on this page. 
6. Stack selected tags horizontally.
7. Add scroll to tags options 
8. Improve mobile view of the page
9. Don't close the tags menu, until user click outside the menu once (enables him to select multiple tags at a time)
10. Improve UI

**A preview of the page -**
practice-section

**Comment by cjchirag7** [comment]
Shall I merge both to master @NBNARADHYA ?

File: server/models/questions/getPublicContestQuestions.js [hunk 2]
```diff
@@ -1,59 +1,133 @@
...
+        difficulty === 'hard')
+    ) {
+      query += `difficulty=? AND `
+      countQuery += `difficulty=? AND `
+      qArr.push(difficulty)
+      cQArr.push(difficulty)
     }
 
-    query += `c.end_time <= NOW() `
+    query += `c.end_time <= NOW() AND c.public  = 1 AND c.confidential_questions = 0 `
```
Comment (cjchirag7) [inline server/models/questions/getPublicContestQuestions.js:93]: Also, for this condition to work correctly, we need to make `confidential_questions` column NOT NULL and with default value 0. I may do this change in the merge itself.

Comment (NBNARADHYA) [reply server/models/questions/getPublicContestQuestions.js:93]: Dont worry about this...I have made changes to server (createContest) to save it accordingly

**Comment by cjchirag7** [comment]
Shifted both changes to #92

**Pull request closed without merging**